.. _pyani-about:

===============
About ``pyani``
===============

``pyani`` is a Python package and standalone program for calculation of whole-genome similarity measures. It is designed to be used with draft or complete prokaryote (bacterial and archaeal) genomes, and implements the following methods:

- ANIb (average nucleotide identity using ``BLAST+``)
- ANIblastall (average nucleotide identity using ``legacy BLAST``)
- ANIm (average nucleotide identity using ``MUMmer``)
- fastANI (average nucleotide identity using ``fastANI``)
- TETRA (4-mer sequence profiles)

-------------------
Funding and Support
-------------------

The development of ``pyani`` has been made possible by generous funding from a number of sources, including:

- The Scottish Government
    - 2011-2016 workpackages
    - 2016-2019 workpackage RD2.1.4

As is the case with much bioinformatics software, ongoing development does not receive continuous direct funding support. If you use ``pyani``, please do cite this tool in your publications and outputs - this helps us justify future funding applications to maintain and improve the software.

You can find more information about how to cite `pyani` on the :ref:`pyani-citations` page.
