.. _pyani-citations:

================
Citing ``pyani``
================

We would be grateful if you could please cite the following manuscript in your work if you have found ``pyani`` useful:

    Pritchard *et al.* (2016) "Genomics and taxonomy in diagnostics for food security: soft-rotting enterobacterial plant pathogens" *Anal. Methods*, 2016, **8**, 12-24 DOI: 10.1039/C5AY02550H

.. code-block:: latex

    @Article{C5AY02550H,
    author ="Pritchard, Leighton and Glover, Rachel H. and Humphris, Sonia and Elphinstone, John G. and Toth, Ian K.",
    title  ="Genomics and taxonomy in diagnostics for food security: soft-rotting enterobacterial plant pathogens",
    journal  ="Anal. Methods",
    year  ="2016",
    volume  ="8",
    issue  ="1",
    pages  ="12-24",
    publisher  ="The Royal Society of Chemistry",
    doi  ="10.1039/C5AY02550H",
    url  ="http://dx.doi.org/10.1039/C5AY02550H",
    abstract  ="Soft rot Enterobacteriaceae (SRE) are bacterial plant pathogens that cause blackleg{,} wilt and soft rot
    diseases on a broad range of important crop and ornamental plants worldwide. These organisms (spanning the genera Erwinia{,}
    Pectobacterium{,} Dickeya{,} and Pantoea) cause significant economic and yield losses in the field{,} and in storage. They
    are transmissible through surface water{,} by trade and other movement of plant material and soil{,} and in some cases are
    subject to international legislative and quarantine restrictions. Effective detection and diagnosis in support of food
    security legislation and epidemiology is dependent on the ability to classify pathogenic isolates precisely. Diagnostics and
    classification are made more difficult by the influence of horizontal gene transfer on phenotype{,} and historically complex
    and sometimes inaccurate nomenclatural and taxonomic assignments that persist in strain collections and online sequence
    databases. Here{,} we briefly discuss the relationship between taxonomy{,} genotype and phenotype in the SRE{,} and their
    implications for diagnostic testing and legislation. We present novel whole-genome classifications of the SRE{,}
    illustrating inconsistencies between the established taxonomies and evidence from completely sequenced isolates. We conclude
    with a perspective on the future impact of widespread whole-genome sequencing and classification methods on detection and
    identification of bacterial plant pathogens in support of legislative and policy efforts in food security."}


.. TIP::
    When citing ``pyani`` in your text, please consider quoting the precise release version (if you are using a specific release), or the specific commit hash (if you are using the 'bleeding-edge' development verison), e.g.

    ``"...using PYANI (v0.3.0)..."``

    or

    ``"...using PYANI (commit bb1cb5c)..."``

    This will enable precise reproduction of this part of your work by others.

-----------------------------
Publications citing ``pyani``
-----------------------------

If you are using ``pyani``, you are in good company. These authors and manuscripts have employed ``pyani`` to help with their whole-genome classification:

(If I've missed out your paper, please do put in a pull request on ``GitHub`` or get in touch and I'll be happy to add it!)

^^^^
2021
^^^^

* Abdel-Glil *et al.* (2021) "Comparative in silico genome analysis of *Clostridium perfringens* unravels stable phylogroups with different genome characteristics and pathogenic potential" *Sci. Rep.* `doi:10.1038/s41598-021-86148-8 <https://doi.org/10.1038/s41598-021-86148-8>`_
* Abdullah *et al.* (2021) "Comparative analysis of whole genome sequences of *Leptospira* spp. from RefSeq database provide interspecific divergence and repertoire of virulence factors" *bioRxiv* `doi:10.1101/2021.01.12.426470 <https://doi.org/10.1101/2021.01.12.426470>`_
* Asselin *et al.* (2021) "Complete genome sequence resources for the onion pathogen, *Pantoea ananatis* OC5a" *Phytopath.* `doi:10.1094/phyto-09-20-0416-a <https://doi.org/10.1094/phyto-09-20-0416-a>`_
* Badhai *et al.* (2021) "Genomic plasticity and antibody response of *Bordetella bronchiseptica* strain HT200, a natural variant from a thermal spring" *FEMS Micro. Lett.* `doi:10.1093/femsle/fnab035 <https://doi.org/10.1093/femsle/fnab035>`_
* Becken *et al.* (2021) "Genotypic and Phenotypic Diversity among Human Isolates of *Akkermansia muciniphila*" *mBio* `doi:10.1128/mBio.00478-21 <https://doi.org/10.1128/mBio.00478-21>`_
* Biffignandi *et al* (2021) "Genome of *Superficieibacter maynardsmithii*, a novel, antibiotic susceptible representative of *Enterobacteriaceae*" *G3* `doi:10.1093/g3journal/jkab019 <https://doi.org/10.1093/g3journal/jkab019>`_
* Boeuf *et al.* (2021) "Meta-pangenomics Reveals Depth-dependent Shifts in Metabolic Potential for the Ubiquitous Marine Bacterial SAR324 Lineage" *Research Squared* `doi:10.21203/rs.3.rs-225427/v1 <https://doi.org/10.21203/rs.3.rs-225427/v1>`_
* Carlin *et al.* (2021) "*Listeria cossartiae sp. nov.*, *Listeria immobilis sp. nov.*, *Listeria portnoyi sp. nov.* and *Listeria rustica sp. nov.*, isolated from agricultural water and natural environments" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.004795 <https://doi.org/10.1099/ijsem.0.004795>`_
* Chew *et al.* (2021) "First isolation of *Candida oceani* from a clinical specimen" *Antonie van Leeuwenhoek* `doi:10.1007/s10482-020-01512-4 <https://doi.org/10.1007/s10482-020-01512-4>`_
* Danneels *et al.* (2021) "Patterns of transmission and horizontal gene transfer in the *Dioscorea sansibarensis* leaf symbiosis revealed by whole-genome sequencing" *Curr. Biol.* `doi:10.1016/j.cub.2021.03.049 <https://doi.org/10.1016/j.cub.2021.03.049>`_
* de Silva *et al.* (2021) "Revisiting the *Colletotrichum* species causing anthracnose of almond in Australia" *Aust. Plant Path.* `doi:10.1007/s13313-020-00765-x <https://doi.org/10.1007/s13313-020-00765-x>`_
* Delgado-Blas *et al.* (2021) "Population genomics and antimicrobial resistance dynamics of *Escherichia coli* in wastewater and river environments" *Commun Biol* `doi:10.1038/s42003-021-01949-x <https://doi.org/10.1038/s42003-021-01949-x>`_
* Ducarmon *et al.* (2021) "Microbiota-associated risk factors for asymptomatic gut colonisation with multi-drug-resistant organisms in a Dutch nursing home" *Genome Medicine* `doi:0.1186/s13073-021-00869-z <https://doi.org/0.1186/s13073-021-00869-z>`_
* Friedrich *et al.* (2021) "Complete Genome Sequence of *Stenotrophomonas indicatrix* DAIF1" *Micro Res. Ann.* `doi:10.1128/MRA.01484-20 <https://doi.org/10.1128/MRA.01484-20>`_
* Friedrich *et al.* (2021) "Living in a Puddle of Mud: Isolation and Characterization of Two Novel *Caulobacteraceae* Strains *Brevundimonas pondensis sp. nov.* and *Brevundimonas goettingensis sp. nov.*" *appl. microbiol.* `doi:10.3390/applmicrobiol1010005 <https://doi.org/10.3390/applmicrobiol1010005>`_
* Gai *et al.* (2021) "Chromosome-scale genome sequence of *Alternaria alternata* causing Alternaria brown spot of citrus" *Mol. Plant Microbe Int.* `doi:10.1094/MPMI-10-20-0278-SC <https://doi.org/10.1094/MPMI-10-20-0278-SC>`_
* Gallardo-Benavente *et al.* (2021) "Genomics Insights into *Pseudomonas* sp. CG01: An Antarctic Cadmium-Resistant Strain Capable of Biosynthesizing CdS Nanoparticles Using Methionine as S-Source" *genes* `doi:10.3390/genes12020187 <https://doi.org/10.3390/genes12020187>`_
* Guerin *et al.* (2021) "Isolation and characterisation of ΦcrAss002, a crAss-like phage from the human gut that infects *Bacteroides xylanisolvens*" *Microbiome* `doi:10.1186/s40168-021-01036-7 <https://doi.org/10.1186/s40168-021-01036-7>`_
* Huang *et al.* (2021) "Phenotypic properties and genotyping analysis of *Bacillus cereus* group isolates from dairy and potato products" *LWT* `doi:10.1016/j.lwt.2021.110853 <https://doi.org/10.1016/j.lwt.2021.110853>`_
* Huang *et al.* (2021) "Genome-resolved metagenomics using environmental and clinical samples" *Brief. Bioinf.* `doi:10.1093/bib/bbab030 <https://doi.org/10.1093/bib/bbab030>`_
* Hünnefeld *et al.* (2021) "Genome Sequence of the Bacteriophage CL31 and Interaction with the Host Strain *Corynebacterium glutamicum* ATCC 13032" *viruses* `doi:10.3390/v13030495 <https://doi.org/10.3390/v13030495>`_
* Ivanova *et al.* (2021) "Draft Genome Assemblies of Two *Campylobacter novaezeelandiae* and Four Unclassified Thermophilic *Campylobacter* Isolates from Canadian Agricultural Surface Water" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00249-21 <https://doi.org/10.1128/MRA.00249-21>`_
* Jian *et al.* (2021) "Diversity and distribution of viruses inhabiting the deepest ocean on Earth" *ISME J.* `doi:10.1038/s41396-021-00994-y <https://doi.org/10.1038/s41396-021-00994-y>`_
* Jungblut *et al.* (2021) "Genomic diversity and CRISPR‐Cas systems in the cyanobacterium *Nostoc* in the High Arctic" *Env. Microbiol.* `doi:10.1111/1462-2920.15481 <https://doi.org/10.1111/1462-2920.15481>`_
* Koirala *et al.* (2021) "Identification of two novel pathovars of *Pantoea stewartii* subsp. *indologenes* affecting Allium sp. and millets" *Phytopathology* `doi:10.1094/PHYTO-11-20-0508-R <https://doi.org/10.1094/PHYTO-11-20-0508-R>`_
* Lee *et al.* (2021) "*Bifidobacterium bifidum* strains synergize with immune checkpoint inhibitors to reduce tumour burden in mice" *Nat. Microbiol.* `doi:10.1038/s41564-020-00831-6 <https://doi.org/10.1038/s41564-020-00831-6>`_
* Lee *et al.* (2021) "Identification and Characterization of a Novel Genomic Island Harboring Cadmium and Arsenic Resistance Genes in *Listeria welshimeri*" *biomolecules* `doi:10.3390/biom11040560 <https://doi.org/10.3390/biom11040560>`_
* Lee *et al.* (2021) "Lemierre’s syndrome associated with hypervirulent *Klebsiella pneumoniae*: A case report and genomic characterization of the isolate" *IDCases* `doi:10.1016/j.idcr.2021.e01173 <https://doi.org/10.1016/j.idcr.2021.e01173>`_
* Li *et al* (2021) "Novel *Paenibacillus sp.* Strains From the Perennial Ryegrass Seed Microbiome Reveal Bioprotectant and Biofertiliser Activity - Differentiating Similar Strains via Genomics and Transcriptomics" *Research Sq.* `doi:10.21203/rs.3.rs-445288/v1 <https://doi.org/10.21203/rs.3.rs-445288/v1>`_
* Liu *et al.* (2021) "*Corynebacterium anserum* sp. nov., isolated from the faeces of greater white-fronted geese (*Anser albifrons*) at Poyang Lake, PR China" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.004637 <https://doi.org/10.1099/ijsem.0.004637>`_
* Lood *et al.* (2021) "Genomics of an endemic cystic fibrosis *Burkholderia multivorans* strain reveals low within-patient evolution but high between-patient diversity" *PLoS Pathog.* `doi:0.1371/journal.ppat.1009418 <https://doi.org/0.1371/journal.ppat.1009418>`_
* Lu *et al.* (2021) "Asgard archaea in the haima cold seep: Spatial distribution and genomic insights" *Deep Sea Res. I* `doi:10.1016/j.dsr.2021.103489 <https://doi.org/10.1016/j.dsr.2021.103489>`_
* Luo *et al.* (2021) "Isolation and characterization of new phage vB_CtuP_A24 and application to control *Cronobacter* spp. in infant milk formula and lettuce" *Food Res. Int.* `doi:10.1016/j.foodres.2021.110109 <https://doi.org/10.1016/j.foodres.2021.110109>`_
* Ma *et al.* (2021) "Identification of *Pectobacterium versatile* causing blackleg of potato in New York State" *Plant Disease* `doi:10.1094/PDIS-09-20-2089-RE <https://doi.oprg/10.1094/PDIS-09-20-2089-RE>`_
* Majer *et al.* (2021) "Whole genome sequencing of *Streptomyces actuosus* ISP-5337, *Streptomyces sioyaensis* B-5408, and *Actinospica acidiphila* B-2296 reveals secondary metabolomes with antibiotic potential" *Biotech. Rep.* `doi:10.1016/j.btre.2021.e00596 <https://doi.org/10.1016/j.btre.2021.e00596>`_
* Matarrita-Carranza *et al.* (2021) "*Streptomyces* sp. M54: an actinobacteria associated with a neotropical social wasp with high potential for antibiotic production." *Antonie van Leeuwenhoek* `doi:10.1007/s10482-021-01520-y <https://doi.org/10.1007/s10482-021-01520-y>`_
* Mao *et al.* (2021) "Comparative Genomic Analysis of *Lactiplantibacillus plantarum* Isolated from Different Niches" *genes* `doi:10.3390/genes12020241 <https://doi.org/10.3390/genes12020241>`_
* Moon *et al* (2021) "Mobile Colistin Resistance Gene mcr-1 Detected on an IncI2 Plasmid in *Salmonella Typhimurium* Sequence Type 19 from a Healthy Pig in South Korea" *microorganisms* `doi:10.3390/microorganisms9020398 <https://doi.org/10.3390/microorganisms9020398>`_
* Moya-Beltrán *et al.* (2021) "Genomic evolution of the class *Acidithiobacillia*: deep-branching *Proteobacteria* living in extreme acidic conditions" *ISME J.* `doi:0.1038/s41396-021-00995-x <https://doi.org/0.1038/s41396-021-00995-x>`_
* Nemec *et al.* (2021) "Delineation of a novel environmental phylogroup of the genus *Acinetobacter* encompassing *Acinetobacter terrae sp. nov.*, *Acinetobacter terrestris sp. nov.* and three other tentative species" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2021.126217 <https://doi.org/10.1016/j.syapm.2021.126217>`_
* Ogg *et al.* (2021) "Pangenome analyses of LuxS-coding genes and enzymatic repertoires in cocoa-related lactic acid bacteria" *Genomics* `doi:10.1016/j.ygeno.2021.04.010 <https://doi.org/10.1016/j.ygeno.2021.04.010>`_
* Öhrman *et al.* (2021) "Reorganized Genomic Taxonomy of *Francisellaceae* Enables Design of Robust Environmental PCR Assays for Detection of *Francisella tularensis*" *Microorganisms* `doi:10.3390/microorganisms9010146 <https://doi.org/10.3390/microorganisms9010146>`_
* Öhrman *et al.* (2021) "Complete Genome Sequence of *Francisella* sp. Strain LA11-2445 (FDC406), a Novel *Francisella* Species Isolated from a Human Skin Lesion" *Micro. Res. Ann.* `doi:10.1128/MRA.01233-20 <https://doi.org/10.1128/MRA.01233-20>`_
* Pais *et al.* (2021) "Genomic sequencing of different sequevars of *Ralstonia solanacearum* belonging to the Moko ecotype" *Genet. Mol. Bol.* `doi:10.1590/1678-4685-gmb-2020-0172 <https://doi.org/10.1590/1678-4685-gmb-2020-0172>`_
* Puri *et al.* "Phylogenomic Framework for Taxonomic Delineation of *Paracoccus* spp. and Exploration of Core-Pan Genome" *Ind. J. Microbiol.* `doi:10.1007/s12088-021-00929-3 <https://doi.org/10.1007/s12088-021-00929-3>`_
* Reichler *et al.* (2021) "Identification, subtyping, and tracking of dairy spoilage-associated *Pseudomonas* by sequencing the *ileS* gene" *J. Dairy Sci.* `doi:10.3168/jds.2020-19283 <https://doi.org/10.3168/jds.2020-19283>`_
* Ryngajłło *et al.* (2021) "Complete genome sequence of lovastatin producer *Aspergillus terreus* ATCC 20542 and evaluation of genomic diversity among *A. terreus* strains" *Appl. Microbiol. Biotechnol.* `doi:10.1007/s00253-021-11133-0 <https://doi.org/10.1007/s00253-021-11133-0>`_
* Schörner *et al.* (2021) "Genomic analysis of *Neisseria elongata* isolate from a patient with infective endocarditis" *FEBS Open Bio* `doi:10.1002/2211-5463.13201 <https://doi.org/10.1002/2211-5463.13201>`_
* Singh *et al.* (2021) "Enrichment and description of novel bacteria performing syntrophic propionate oxidation at high ammonia level" *Env. Micro.* `doi:10.1111/1462-2920.15388 <https://doi.org/10.1111/1462-2920.15388>`_
* Singh *et al.* (2021) "Genome-based reclassification of *Amycolatopsis eurytherma* as a later heterotypic synonym of *Amycolatopsis thermoflava*" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.004642 <https://doi.org/10.1099/ijsem.0.004642>`_
* Son *et al.* (2021) "*Serratia rhizosphaerae sp. nov.*, a novel plant resistance inducer against soft rot disease in tobacco" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.004788 <https://doi.org/10.1099/ijsem.0.004788>`_
* Sorokin *et al.* (2021) "*Natronoglycomyces albus gen. nov., sp. nov*, a haloalkaliphilic actinobacterium from a soda solonchak soil" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.004804 <https://doi.org/10.1099/ijsem.0.004804>`_
* Suarez *et al.* (2021) "Whole-Genome sequencing and comparative genomics of *Mycobacterium spp.* from farmed Atlantic and coho salmon in Chile" *Antonie van Leeuw.* `doi:10.1007/s10482-021-01592-w <https://doi.org/10.1007/s10482-021-01592-w>`_
* Tian *et al.* (2021) "LINflow: a computational pipeline that combines an alignment-free with an alignment-based method to accelerate generation of similarity matrices for prokaryotic genomes" *PeerJ* `doi:10.7717/peerj.10906 <https://doi.org/10.7717/peerj.10906>`_
* Tian *et al.* (2021) "Antifungal mechanism of *Bacillus amyloliquefaciens* strain GKT04 against *Fusarium* wilt revealed using genomic and transcriptomic analyses" *Microbiol. Open* `doi:10.1002/mbo3.1192 <https://doi.org/10.1002/mbo3.1192>`_
* van der Lelie *et al.* (2021) "Rationally designed bacterial consortia to treat chronic immune-mediated colitis and restore intestinal homeostasis" *Nat. Comms.* `doi:10.1038/s41467-021-23460-x <https://doi.org/10.1038/s41467-021-23460-x>`_
* Vargas-Peralta *et al.* (2021) "Identification of *Pectobacterium* species isolated from the soft rot of tetecho (*Neobuxbaumia tetetzo*), a columnar cactus, and associated metagenomics" *bioRxiv* `doi:10.1101/2021.02.01.429127 <https://doi.org/10.1101/2021.02.01.429127>`_
* Verma *et al.* (2021) "Genome analyses of 174 strains of *Mycobacterium tuberculosis* provide insight into the evolution of drug resistance and reveal potential drug targets" *Microb. Genom.* `doi:10.1099/mgen.0.000542 <https://doi.org/10.1099/mgen.0.000542>`_
* Vincent *et al.* (2021) "AsaGEI2d: a new variant of a genomic island identified in a group of *Aeromonas salmonicida* subsp. *salmonicida* isolated from France, which bears the pAsa7 plasmid" *FEMS Micro. Lett.* `doi:10.1093/femsle/fnab021 <https://doi.org/10.1093/femsle/fnab021>`_
* Volpiano *et al.* (2021) "Genomic Metrics Applied to *Rhizobiales (Hyphomicrobiales)*: Species Reclassification, Identification of Unauthentic Genomes and False Type Strains" *Front. Microbiol.* `doi:10.3389/fmicb.2021.614957 <https://doi.org/10.3389/fmicb.2021.614957>`_
* Watson *et al.* (2021) "Adaptive ecological processes and metabolic independence drive microbial colonization and resilience in the human gut" *bioRxiv* `doi:10.1101/2021.03.02.433653 <https://doi.org/10.1101/2021.03.02.433653>`_
* Wu *et al.* (2021) "Metagenomic insights into nitrogen and phosphorus cycling at the soil aggregate scale driven by organic material amendments" *Sci. Tot. Env.* `doi:10.1016/j.scitotenv.2021.147329 <https://doi.org/10.1016/j.scitotenv.2021.147329>`_
* Young *et al.* (2021) "Defining the *Rhizobium leguminosarum* Species Complex" *genes* `doi:10.3390/genes12010111 <https://doi.org/10.3390/genes12010111>`_
* Zeng *et al.* (2021) "Novel phage vB_CtuP_B1 for controlling *Cronobacter malonaticus* and *Cronobacter turicensis* in ready-to-eat lettuce and powered infant formula" *Food Res. Int.* `doi:10.1016/j.foodres.2021.110255 <https://doi.org/10.1016/j.foodres.2021.110255>`_
* Zvi-Kedem *et al.* (2021) "The worm affair: Genetic diversity in two species of symbionts that co-occur in tubeworms from the Mediterranean Sea" *bioRxiv* `doi:10.1101/2021.01.27.428081 <https://doi.org/10.1101/2021.01.27.428081>`_

^^^^
2020
^^^^

* Akob *et al.* (2020) "Mixotrophic Iron-Oxidizing *Thiomonas* Isolates from an Acid Mine Drainage-Affected Creek" *App. Env. Microbiol.* `doi:10.1128/AEM.01424-20 <https://doi.org/10.1128/AEM.01424-20>`_
* Albert *et al.* (2020) "Comparative Pangenomics of the Mammalian Gut Commensal *Bifidobacterium longum*" *Microorganisms* `doi:10.3390/microorganisms8010007 <https://doi.org/10.3390/microorganisms8010007>`_
* Bech *et al.* (2020) "Marine Sediments Hold an Untapped Potential for Novel Taxonomic and Bioactive Bacterial Diversity" *mSystems* `doi:/10.1128/mSystems.00782-20 <https://doi.org//10.1128/mSystems.00782-20>`_
* Biggel *et al.* (2020) "Horizontally acquired papGII-containing pathogenicity islands underlie the emergence of invasive uropathogenic *Escherichia coli* lineages" *Nat, Comm.* `doi:10.1038/s41467-020-19714-9 <https://doi.org/10.1038/s41467-020-19714-9>`_
* Bisanz *et al.* (2020) "A Genomic Toolkit for the Mechanistic Dissection of Intractable Human Gut Bacteria" *Cell Host & Microbe* `doi:10.1016/j.chom.2020.04.006 <https://doi.org/10.1016/j.chom.2020.04.006>`_
* Bridel *et al.* (2020) "Genetic diversity and population structure of *Tenacibaculum maritimum*, a serious bacterial pathogen of marine fish: from genome comparisons to high throughput MALDI-TOF typing" *Vet. Res.* `doi:10.1186/s13567-020-00782-0 <https://doi.org/10.1186/s13567-020-00782-0>`_
* Brock *et al.* (2020) "Endosymbiotic adaptations in three new bacterial species associated with *Dictyostelium discoideum*: *Paraburkholderia agricolaris* sp. nov., *Paraburkholderia hayleyella* sp. nov., and *Paraburkholderia bonniea* sp. nov" *PeerJ* `doi:10.7717/peerj.9151 <https://doi.org/10.7717/peerj.9151>`_
* Busch *et al.* (2020) "Using affinity propagation clustering for identifying bacterial clades and subclades with whole-genome sequences of *Francisella tularensis*" *PLoS Neg. Trop. Dis.* `doi:10.1371/journal.pntd.0008018 <https://doi.org/10.1371/journal.pntd.0008018>`_
* Cai *et al.* (2020) "Comparative genomics of *Klebsiella michiganensis* BD177 and related members of *Klebsiella* sp. reveal the symbiotic relationship with *Bactrocera dorsalis*" *BMC Genetics* `doi:10.1186/s12863-020-00945-0 <https://doi.org/10.1186/s12863-020-00945-0>`_
* Cassaniti *et al.* (2020) "Authors' response: COVID-19: how accurate are seroprevalence studies?" *Eurosurveillance* `doi:10.2807/1560-7917.ES.2020.25.30.2001437 <https://doi.org/10.2807/1560-7917.ES.2020.25.30.2001437>`_
* Chibani *et al.* (2020) "Genomic variation among closely related *Vibrio alginolyticus* strains is located on mobile genetic elements" *BMC Genomics* `doi:10.1186/s12864-020-6735-5 <https://doi.org/10.1186/s12864-020-6735-5>`_
* Christman *et al.* (2020) "Novel clostridial lineages recovered from metagenomes of a hot oil reservoir" *Sci. Rep.* `doi:10.1038/s41598-020-64904-6 <https://doi.org/10.1038/s41598-020-64904-6>`_
* Christman *et al.* (2020) "Methanogens Within a High Salinity Oil Reservoir From the Gulf of Mexico" *Front. Microbiol.* `doi:10.3389/fmicb.2020.570714 <https://doi.org/10.3389/fmicb.2020.570714>`_
* Cunningham-Oakes *et al.* (2020) "Genome Sequence of *Pluralibacter gergoviae* ECO77, a Multireplicon Isolate of Industrial Origin" *Microbiol. Res. Ann.* `doi:10.1128/MRA.01561-19 <https://doi.org/0.1128/MRA.01561-19>`_
* Dahihausen *et al.* (2020) "Isolation and sequence-based characterization of a koala symbiont: *Lonepinella koalarum*" *PeerJ* `doi:10.7717/peerj.10177 <https://doi.org/10.7717/peerj.10177>`_
* Dam *et al.* (2020) "Targeted Cell Sorting Combined With Single Cell Genomics Captures Low Abundant Microbial Dark Matter With Higher Sensitivity Than Metagenomics" *Front. Microbiol.* `doi:10.3389/fmicb.2020.01377 <https://doi.org/10.3389/fmicb.2020.01377>`_
* Damnjanovich *et al.* (2020) "Bacteriophage genotyping using BOXA repetitive-PCR" *BMC Microbiol.* `doi:10.1186/s12866-020-01770-2 <https://doi.org/10.1186/s12866-020-01770-2>`_
* Dangel *et al.* (2020) "*Corynebacterium silvaticum* sp. nov., a unique group of NTTB corynebacteria in wild boar and roe deer" *Int. J. Syst. Evol. Microb.* `doi:10.1099/ijsem.0.004195 <https://doi.org/10.1099/ijsem.0.004195>`_
* de Andrade Alves *et al.* "First isolation and whole-genome sequencing of a *Shewanella algae* strain from a swine farm in Brazil" *BMC Microbiol.* `doi:10.1186/s12866-020-02040-x <https://doi.org/10.1186/s12866-020-02040-x>`_
* Déraspe *et al.* (2020) "Genome Sequence of a *Klebsiella pneumoniae* NDM-1 Producer Isolated in Quebec City" *Microbiol. Resour. Announc.* `doi:10.1128/MRA.00829-19 <https://doi.org/10.1128/MRA.00829-19>`_
* Donner *et al.* (2020) "Septic shock caused by *Capnocytophaga canis* after a cat scratch" *Eur. J. Clin. Microbiol. Inf. Dis.* `doi:10.1007/s10096-020-03922-8 <https://doi.org/10.1007/s10096-020-03922-8>`_
* Duar *et al.* (2020) "Comparative Genome Analysis of *Bifidobacterium longum* subsp. *infantis* Strains Reveals Variation in Human Milk Oligosaccharide Utilization Genes among Commercial Probiotics" *Nutrients* `doi:10.3390/nu12113247 <https://doi.org/10.3390/nu12113247>`_
* Elcheninov *et al.* (2020) "*Thermogemmata fonticola* gen. nov., sp. nov., the first thermophilic planctomycete of the order *Gemmatales* from a Kamchatka hot spring" *Syst. App. Micro.* `doi:10.1016/j.syapm.2020.126157 <https://doi.org/10.1016/j.syapm.2020.126157>`_
* Feng *et al.* (2020) "Phylogenetic and genomic analysis reveals high genomic openness and genetic diversity of *Clostridium perfringens*" *Microb. Gen.* `doi:10.1099/mgen.0.000441 <https://doi.org/10.1099/mgen.0.000441>`_
* Ferrerira *et al.* "Genome-based reclassification of *Azospirillum brasilense* Sp245 as the type strain of *Azospirillum baldaniorum* sp. nov" *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.004517 <https://doi.org/10.1099/ijsem.0.004517>`_
* Fishbein *et al.* (2020) "Randomized Controlled Trial of Oral Vancomycin Treatment in *Clostridioides difficile*-Colonized Patients" *mSphere* `doi:10.1128/mSphere.01296-20 <https://doi.org/10.1128/mSphere.01296-20>`_
* France *et al.* (2020) "Complete Genome Sequences of Six *Lactobacillus iners* Strains Isolated from the Human Vagina" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00234-20 <https://doi.org/10.1128/MRA.00234-20>`_
* Francoeur *et al.* (2020) "Bacteria Contribute to Plant Secondary Compound Degradation in a Generalist Herbivore System" *mBio* `doi:doi.org/10.1128/mBio.02146-20 <https://doi.org/doi:doi.org/10.1128/mBio.02146-20>`_
* Freitas *et al.* (2020) "Yeast communities associated with cacti in Brazil and the description of *Kluyveromyces starmeri* sp. nov. based on phylogenomic analyses" *Yeast* `doi:10.1002/yea.3528 <https://doi.org/10.1002/yea.3528>`_
* Friedrich *et al.* (2020) "First Complete Genome Sequences of *Janthinobacterium lividum* EIF1 and EIF2 and their Comparative Genome Analysis" *Genome Biol. Evol.* `doi:10.1093/gbe/evaa148 <https://doi.org/10.1093/gbe/evaa148>`_
* Furrer *et al.* (2020) "Phage vB_BveM-Goe7 represents a new genus in the subfamily *Bastillevirinae*" *Arch. Virol.* `doi:10.1007/s00705-020-04546-1 <https://doi.org/10.1007/s00705-020-04546-1>`_
* Gabor *et al.* (2020) "A New Species of Genus *Pseudomonas*" United States Patent Application 20200216503 `20200216503 <http://www.freepatentsonline.com/y2020/0216503.html>`_
* Gai *et al.* (2020) "The Genome Sequence of the Citrus Melanose Pathogen *Diaporthe citri* and Two Citrus related *Diaporthe* species" *Phytopathology* `doi:10.1094/PHYTO-08-20-0376-SC <https://doi.org/10.1094/PHYTO-08-20-0376-SC>`_
* Gardon *et al.* (2020) "A drift‐barrier model drives the genomic landscape of a structured bacterial population" *Molecular Ecol.* `doi:10.1111/mec.15628 <https://doi.org/10.1111/mec.15628>`_
* Girard *et al.* (2020) "Reliable Identification of Environmental *Pseudomonas* Isolates Using the *rpoD* Gene" *Microorganisms* `doi:10.3390/microorganisms8081166 <https://doi.org/10.3390/microorganisms8081166>`_
* González-Dominici *et al.* (2020) "Genome Analysis and Genomic Comparison of the Novel Species *Arthrobacter ipsi* Reveal Its Potential Protective Role in Its Bark Beetle Host" *Microbial Ecol.* `doi:10.1007/s00248-020-01593-8 <https://doi.org/10.1007/s00248-020-01593-8>`_
* González-Gómez *et al.* (2020) "Phylogenomic Analysis Supports Two Possible Origins for Latin American Strains of *Vibrio parahaemolyticus* Associated with Acute Hepatopancreatic Necrosis Disease (AHPND)" *Curr. Microbiol.* `doi:10.1007/s00284-020-02214-w <https://doi.org/10.1007/s00284-020-02214-w>`_
* Gramaje *et al.* (2020) "Comparative Genomic Analysis of *Dactylonectria torresensis* Strains from Grapevine, Soil and Weed Highlights Potential Mechanisms in Pathogenicity and Endophytic Lifestyle" *J. Fungi* `doi:10.3390/jof6040255 <https://doi.org/10.3390/jof6040255>`_
* Graña-Miraglia *et al.* (2020) "Spirochetes isolated from arthropods constitute a novel genus *Entomospira* genus novum within the order Spirochaetales" *Sci. Rep.* `doi:10.1038/s41598-020-74033-9 <https://doi.org/10.1038/s41598-020-74033-9>`_
* Hempel *et al.* (2020) "Complete Genome Sequence of *Bacillus velezensis* Strain S4, Isolated from Biochar-Treated Soil" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00352-20 <https://doi.org/10.1128/MRA.00352-20>`_
* Hinger *et al.* (2020) "Phylogenomic Analyses of Members of the Widespread Marine Heterotrophic Genus *Pseudovibrio* Suggest Distinct Evolutionary Trajectories and a Novel Genus, *Polycladidibacter* gen. nov." *Appl. Env. Microbiol.* `doi:10.1128/AEM.02395-19 <https://doi.org/10.1128/AEM.02395-19>`_
* Hollensteiner *et al.* (2020) "Genome Sequence of *Komagataeibacter saccharivorans* Strain JH1, Isolated from Fruit Flies" *Microbiol. Res. Announc.* `doi:10.1128/MRA.00098-20 <https://doi.org/10.1128/MRA.00098-20>`_
* Hulin *et al.* (2020) "Cherry picking by pseudomonads: after a century of research on canker, genomics provides insights into the evolution of pathogenicity towards stone fruits" *Plant Pathology* `doi:10.1111/ppa.13189 <https://doi.org/10.1111/ppa.13189>`_
* Ibarra Caballero *et al.* (2020) "Genome comparison and transcriptome analysis of the invasive brown root rot pathogen, *Phellinus noxius*, from different geographic regions reveals potential enzymes associated with degradation of different wood substrates" *Fungal Biology* `doi:10.1016/j.funbio.2019.12.007 <https://doi.org/10.1016/j.funbio.2019.12.007>`_
* Inderbitzin *et al.* (2020) "Species identification in plant-associated prokaryotes and fungi using DNA" *Phytobiomes J.* `doi:10.1094/PBIOMES-12-19-0067-RVW <https://doi.org/10.1094/PBIOMES-12-19-0067-RVW">`_
* Jin *et al* (2020) "Complete genome sequence of fish‑pathogenic *Aeromonas hydrophila* HX‑3 and a comparative analysis: insights into virulence factors and quorum sensing" *Sci. Rep.* `doi:10.1038/s41598-020-72484-8 <https://doi.org/10.1038/s41598-020-72484-8>`_
* Joglekar *et al.* (2020) "Polyphasic analysis reveals correlation between phenotypic and genotypic analysis in soybean bradyrhizobia (*Bradyrhizobium* spp.)" *Syst. Appl. Microb.* `doi:10.1016/j.syapm.2020.126073 <https://doi.org/10.1016/j.syapm.2020.126073>`_
* Joutsen *et al.* (2020) "Two copies of the *ail* gene found in *Yersinia enterocolitica* and *Yersinia kristensenii*" *Vet. Micro.* `doi:10.1016/j.vetmic.2020.108798 <https://doi.org/10.1016/j.vetmic.2020.108798>`_
* Jung *et al.* (2020) "Genome Analysis of *Enterococcus mundtii* Pe103, a Human Gut-Originated Pectinolytic Bacterium" *Curr. Microbiol.* `doi:10.1007/s00284-020-01932-5 <https://doi.org/10.1007/s00284-020-01932-5>`_
* Kim *et al.* (2020) "Genome analysis of *Lactobacillus plantarum* subsp. *plantarum* KCCP11226 reveals a well-conserved C30 carotenoid biosynthetic pathway" *3 Biotech.* `doi:10.1007/s13205-020-2149-y <https://doi.org/10.1007/s13205-020-2149-y>`_
* Kim *et al.* (2020) "Comparative Genomics Determines Strain-Dependent Secondary Metabolite Production in *Streptomyces venezuelae* Strains" *Biomolecules* `doi:10.3390/biom10060864 <https://doi.org/10.3390/biom10060864>`_
* Kornienko *et al.* (2020) "Contribution of *Podoviridae* and *Myoviridae* bacteriophages to the effectiveness of anti-staphylococcal therapeutic cocktails" *Sci. Rep.* `doi:10.1038/s41598-020-75637-x <https://doi.org/10.1038/s41598-020-75637-x>`_
* Kroll *et al.* (2020) "Microbiota supplementation with *Bifidobacterium* and *Lactobacillus* modifies the preterm infant gut microbiota and metabolome: an observational study" *Cell Reports Medicine* `doi:10.1016/j.xcrm.2020.100077 <https://doi.org/10.1016/j.xcrm.2020.100077>`_
* Kuleshov *et al.* (2020) "Whole genome sequencing of *Borrelia miyamotoi* isolate Izh-4: reference for a complex bacterial genome" *BMC Genomics* `doi:10.1186/s12864-019-6388-4 <https://doi.org/10.1186/s12864-019-6388-4>`_
* Kumar *et al.* (2020) "Phylogenetic Relationship Among Brackishwater *Vibrio* Species" *Evol. Bioinf.* `doi:10.1177/1176934320903288 <https://doi.org/10.1177/1176934320903288>`_
* La China *et al.* (2020) "Genome sequencing and phylogenetic analysis of K1G4: a new *Komagataeibacter* strain producing bacterial cellulose from different carbon sources" *Biotech. Lett.* `doi:10.1007/s10529-020-02811-6 <https://doi.org/10.1007/s10529-020-02811-6>`_
* Lacault *et al.* (2020) "Zucchini vein clearing disease is caused by several lineages within *Pseudomonas syringae* species complex." *Phytopathology* `doi:10.1094/PHYTO-07-19-0266-R <https://doi.org/10.1094/PHYTO-07-19-0266-R>`_
* Leyer *et al.* (2020) "*Avrilella dinanensis* gen. nov., sp. nov., a novel bacterium of the family *Flavobacteriaceae* isolated from human blood" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2020.126124 <https://doi.org/10.1016/j.syapm.2020.126124>`_
* Li *et al.* (2020) "Metabolic diversification of anaerobic methanotrophic archaea in a deep-sea cold seep" *Marine Life Sci. Tech.* `doi:10.1007/s42995-020-00057-9 <https://doi.org/10.1007/s42995-020-00057-9>`_
* Li *et al.* (2020) "Comparative Genomics Reveals Broad Genetic Diversity, Extensive Recombination and Nascent Ecological Adaptation in *Micrococcus luteus*" *Research Square* `doi:10.21203/rs.3.rs-62334/v1 <https://doi.org/10.21203/rs.3.rs-62334/v1>`_
* Li *et al.* (2020) "Isolation and Characterization of *Bacillus cereus* Phage vB_BceP-DLc1 Reveals the Largest Member of the Phi29-Like Phages" *Microorganisms* `doi:10.3390/microorganisms8111750 <https://doi.org/:10.3390/microorganisms8111750>`_
* Liu *et al.* (2020) "Whole genome sequence and comparative genome analyses of multi-resistant *Staphylococcus warneri* GD01 isolated from a diseased pig in China" *PLoS One* `doi:10.1371/journal.pone.0233363 <https://doi.org/10.1371/journal.pone.0233363>`_
* Liu *et al.* (2020) "Pathogenicity of a *Vibrio owensii* strain isolated from *Fenneropenaeus chinensis* carrying pirAB genes and causing AHPND" *Aquaculture* `doi:10.1016/j.aquaculture.2020.735747 <https://doi.org/10.1016/j.aquaculture.2020.735747>`_
* Long *et al.* (2020) "Polyclonality, Shared Strains, and Convergent Evolution in Chronic CF *S. aureus* Airway Infection" *Am. J. Resp. Crit. Care Med.* `doi:10.1164/rccm.202003-0735O <https://doi.org/10.1164/rccm.202003-0735OC>`_
* Machado *et al.* (2020) "Molecular relationships of *Campomanesia xanthocarpa* within Myrtaceae based on the complete plastome sequence and on the plastid ycf2 gene" *Genet. Mol. Biol.* `doi:10.1590/1678-4685-gmb-2018-0377  <https://doi.org/10.1590/1678-4685-gmb-2018-0377>`_
* Méndez *et al.* (2020) "Comparative Genomics of Pathogenic *Clavibacter michiganensis* subsp. *michiganensis* Strains from Chile Reveals Potential Virulence Features for Tomato Plants" *Microorganisms* `doi:10.3390/microorganisms8111679 <https://doi.org/0.3390/microorganisms8111679>`_
* Min *et al.* (2020) "Complete Genomic Analysis of *Enterococcus faecium* Heat-Resistant Strain Developed by Two-Step Adaptation Laboratory Evolution Method" *Front. Bioeng. Biotechnol.* `doi:10.3389/fbioe.2020.00828 <https://doi.org/10.3389/fbioe.2020.00828>`_
* Mino *et al.* (2020) "*Hydrogenimonas urashimensis* sp. nov., a hydrogen-oxidizing chemolithoautotroph isolated from a deep-sea hydrothermal vent in the Southern Mariana Trough" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2020.126170 <https://doi.org/10.1016/j.syapm.2020.126170>`_
* Miyoshi *et al.* (2020) "Early-life microbial intervention reduces colitis risk promoted by antibiotic-induced gut dysbiosis" *bioRxiv* `doi:10.1101/2020.03.11.987412 <https://doi.org/10.1101/2020.03.11.987412>`_
* Modesto *et al.* (2020) "Bifidobacteria in two-toed sloths (*Choloepus didactylus*): phylogenetic characterization of the novel taxon *Bifidobacterium choloepi* sp. nov." *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.004506 <https://doi.org/10.1099/ijsem.0.004506>`_
* Moon *et al.* (2020) "First Report of an *Escherichia coli* Strain Carrying the Colistin Resistance Determinant *mcr-1* from a Dog in South Korea" *Antibiotics* `doi:10.3390/antibiotics9110768 <https://doi.org/10.3390/antibiotics9110768>`_
* Moore *et al.* (2020) "Draft Genome Sequence of *Lactobacillus rhamnosus* NCB 441, Isolated from Egyptian White Domiati Cheese" *Micro. Res. Ann.* `doi:10.1128/MRA.01191-20 <https://doi.org/10.1128/MRA.01191-20>`_
* Mu *et al.* (2020) "*Bradymonabacteria*, a novel bacterial predator with versatile survival strategies in saline environments" *Microbiome* `doi:10.21203/rs.2.20535/v1 <https://doi.org/10.21203/rs.2.20535/v1>`_
* Mu *et al.* (2020) "*Tichowtungia aerotolerans* gen. nov., sp. nov., a novel representative of the phylum *Kiritimatiellaeota* and proposal of *Tichowtungiaceae* fam. nov., *Tichowtungiales* ord. nov. and *Tichowtungiia* class. nov." *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.004370 <https://doi.org/10.1099/ijsem.0.004370>`_
* Müller *et al.* (2020) "*Aliarcobacter butzleri* from Water Poultry: Insights into Antimicrobial Resistance, Virulence and Heavy Metal Resistance" *Genes* `doi:10.3390/genes11091104 <https://doi.org/10.3390/genes11091104>`_
* Mullins *et al.* (2020) "Genomic Assemblies of Members of *Burkholderia* and Related Genera as a Resource for Natural Product Discovery" *Micro. Res. Ann.* `doi:10.1128/MRA.00485-20 <https://doi.org/10.1128/MRA.00485-20>`_
* Mullins *et al.* (2020) "Reclassification of the biocontrol agents *Bacillus subtilis* BY-2 and Tu-100 as *Bacillus velezensis* and insights into the genomic and specialized metabolite diversity of the species" *Microbiol.* `doi:10.1099/mic.0.000986 <https://doi.org/10.1099/mic.0.000986>`_
* Nilsson *et al.* (2020) "Diversity and Host Interactions Among Virulent and Temperate Baltic Sea *Flavobacterium* Phages" *Viruses* `doi:10.3390/v12020158 <https://doi.org/10.3390/v12020158>`_
* Norris *et al.* (2020) "*Acidithiobacillus ferrianus* sp. nov.: an ancestral extremely acidophilic and facultatively anaerobic chemolithoautotroph" *Extremophiles* `doi:0.1007/s00792-020-01157-1 <https://doi.org/0.1007/s00792-020-01157-1>`_
* Oshkin *et al.* (2020) "Pan-Genome-Based Analysis as a Framework for Demarcating Two Closely Related Methanotroph Genera Methylocystis and Methylosinus" *Microorganisms* `doi:10.3390/microorganisms8050768 <https://doi.org/10.3390/microorganisms8050768>`_
* Pandey *et al.* (2020) "Evidence of homologous recombination as a driver of diversity in *Brachyspira pilosicoli*" *Micro. Genom.* `doi:10.1099/mgen.0.000470 <https://doi.org/10.1099/mgen.0.000470>`_
* Paulsen *et al.* (2020) "Production of the antimicrobial compound tetrabromopyrrol and the *Pseudomonas* quinolone system precursor, 2‑heptyl‑4‑quinolone, by a novel marine species *Pseudoalteromonas galatheae* sp. nov." *Sci. Rep.* `doi:10.1038/s41598-020-78439-3 <https://doi.org/10.1038/s41598-020-78439-3>`_
* Pang *et al.* (2020) "The Genomic Context for the Evolution and Transmission of Community-Associated *Staphylococcus aureus* ST59 Through the Food Chain" *Front. Microbiol.* `doi:10.3389/fmicb.2020.00422 <https://doi.org/10.3389/fmicb.2020.00422>`_
* Panwar *et al.* (2020) "Influence of the polar light cycle on seasonal dynamics of an Antarctic lake microbial community" *Microbiome* `doi:10.1186/s40168-020-00889-8 <https://doi.org/10.1186/s40168-020-00889-8>`_
* Parlikar *et al.* (2020) "Understanding genomic diversity, pan-genome, and evolution of SARS-CoV-2" *PeerJ* `doi:10.7717/peerj.9576 <https://doi.org/10.7717/peerj.9576>`_
* Pasanen *et al.* (2020) "*Pectobacterium parvum* sp. nov., having a Salmonella SPI-1-like Type III secretion system and low virulence" *Int. J. Syst. Evol. Microb.* `doi:10.1099/ijsem.0.004057 <https://doi.org/10.1099/ijsem.0.004057>`_
* Peral-Aranega *et al.* (2020) "Bacteria Belonging to *Pseudomonas typographi* sp. nov. from the Bark Beetle *Ips typographus* Have Genomic Potential to Aid in the Host Ecology" *Insect* `doi::10.3390/insects11090593 <https://doi.org/:10.3390/insects11090593>`_
* Pierry *et al.* (2020) "High-quality draft genome sequence resources of eight *Xylella fastidiosa* strains isolated from citrus, coffee, plum and hibiscus in South America" *Phytopathology* `doi:10.1094/PHYTO-05-20-0162-A <https://doi.org/10.1094/PHYTO-05-20-0162-A>`_
* Pierry *et al.* (2020) "Genetic Diversity of *Xylella fastidiosa* Plasmids Assessed by Comparative Genomics" *Trop. Plant Path.* `doi:doi.org/10.1007/s40858-020-00359-4 <https://doi.org/doi.org/10.1007/s40858-020-00359-4>`_
* Portier *et al.* (2020) "Updated taxonomy of *Pectobacterium* genus in the CIRM-CFBP bacterial collection: when newly described species reveal “old” endemic population" *Preprints* `doi:10.20944/preprints202008.0608.v1 <https://doi.org/10.20944/preprints202008.0608.v1>`_
* Rackaityte (2020) "Viable bacterial colonization is highly limited in the human intestine in utero" *Nature Medicine* `doi:10.1038/s41591-020-0761-3 <https://doi.org/10.1038/s41591-020-0761-3>`_
* Roach *et al.* (2020) "Whole genome sequencing of Peruvian *Klebsiella pneumoniae* identifies novel plasmid vectors bearing carbapenem resistance gene NDM-1" *Open Forum Inf. Dis.* `doi:10.1093/ofid/ofaa266/5866602 <https://doi.org/10.1093/ofid/ofaa266/5866602>`_
* Rothen *et al.* (2020) "A simple, rapid typing method for *Streptococcus agalactiae* based on ribosomal subunit proteins by MALDI-TOF MS" *Sci. Reports* `doi:10.1038/s41598-020-65707-5 <https://doi.org/10.1038/s41598-020-65707-5>`_
* Ryngajłło *et al.* (2020) "Towards control of cellulose biosynthesis by *Komagataeibacter* using systems-level and strain engineering strategies: current progress and perspectives" *Appl. Microbil. Biotech.* `doi:10.1007/s00253-020-10671-3 <https://doi.org/10.1007/s00253-020-10671-3>`_
* Salgar-Chaparro *et al.* (2020) "Complete Genome Sequence of *Pseudomonas balearica* Strain EC28, an Iron-Oxidizing Bacterium Isolated from Corroded Steel" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00275-20 <https://doi.org/10.1128/MRA.00275-20>`_
* Salgar-Chaparro *et al.* (2020) " Draft Genome Sequence of *Enterobacter roggenkampii* Strain OS53, Isolated from Corroded Pipework at an Offshore Oil Production Facility" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00583-20 <https://doi.org/10.1128/MRA.00583-20>`_
* Salgar-Chaparro *et al.* (2020) "Complete Genome Sequence of *Shewanella chilikensis* Strain DC57, Isolated from Corroded Seal Rings at a Floating Oil Production System in Australia" *Microbiol. Res. Announc.* `doi:0.1128/MRA.00584-20 <https://doi.org/0.1128/MRA.00584-20>`_
* Shen *et al.* (2020) "*Helicobacter monodelphidis* sp. nov. and *Helicobacter didelphidarum* sp. nov., isolated from grey short-tailed opossums (*Monodelphis domestica*) with endemic cloacal prolapses" *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.004424 <https://doi.org/10.1099/ijsem.0.004424>`_
* Strang (2020) "Genomic Insights and Ecological Adaptations of Deep-Subsurface and Near Subsurface *Thermococcus* Isolates and Near Subsurface *Thermococcus* Isolates" *WWU Graduate School Collection* `https://cedar.wwu.edu/wwuet/926 <https://cedar.wwu.edu/wwuet/926>`_
* Taparia *et al.* (2020) "Molecular characterization of *Pseudomonas* from *Agaricus bisporus* caps reveal novel blotch pathogens in Western Europe" *BMC Genomics* `doi:10.1186/s12864-020-06905-3 <https://doi.org/10.1186/s12864-020-06905-3>`_
* Tardy *et al.* (2020) "*Mycoplasma bovis* in Nordic European Countries: Emergence and Dominance of a New Clone" *Pathogens* `doi:10.3390/pathogens9110875 <https://doi.org/10.3390/pathogens9110875>`_
* Thapa *et al.* (2020) "Genome‐wide analyses of Liberibacter species provides insights into evolution, phylogenetic relationships, and virulence factors" *Mol. Plant Path.* `doi:10.1111/mpp.12925  <https://doi.org/10.1111/mpp.12925>`_
* Tian *et al.* (2020) "LINbase: a web server for genome-based identification of prokaryotes as members of crowdsourced taxa" *Nuc. Acids Res.* `doi:10.1093/nar/gkaa190 <https://doi.org/10.1093/nar/gkaa190>`_
* Tsukimi *et al.* (2020) "Draft Genome Sequences of *Bifidobacterium animalis* Consecutively Isolated from Healthy Japanese Individuals" *J. Genomics* `doi:10.7150/jgen.38516 <https://doi.org/10.7150/jgen.38516>`_
* Vijayan *et al.* (2020) "Bacteria known to induce settlement of larvae of *Hydroides elegans* are rare in natural inductive biofilm" *Aquatic Microb. Ecol.* `doi:10.3354/ame01925 <https://doi.org/10.3354/ame01925>`_
* Waleron *et al.* (2020) "*Arthrospiribacter ruber* gen. nov., sp. nov., a novel bacterium isolated from *Arthrospira* cultures" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2020.126072 <https://doi.org/10.1016/j.syapm.2020.126072>`_
* Wang *et al.* (2020) "Comparative Genomics Analysis of *Lactobacillus ruminis* from Different Niches" *Genes* `doi:10.3390/genes11010070 <https://doi.org/10.3390/genes11010070>`_
* Wang *et al.* (2020) "Complete genomic data of Burkholderia glumae strain GX associated with bacterial panicle blight of rice in China" *Plant Dis.* `doi:10.1094/PDIS-10-19-2265-A <https://doi.org/10.1094/PDIS-10-19-2265-A>`_
* Weiser *et al.* (2020) "A Novel Inducible Prophage from *Burkholderia Vietnamiensis* G4 is Widely Distributed across the Species and Has Lytic Activity against Pathogenic Burkholderia" *Viruses* `doi:10.3390/v12060601 <https://doi.org/10.3390/v12060601>`_
* Webster *et al.* (2020) "Culturable diversity of bacterial endophytes associated with medicinal plants of the Western Ghats, India" *FEMS Microbiol. Ecol.* `doi:10.1093/femsec/fiaa147/5876344 <https://doi.org/10.1093/femsec/fiaa147/5876344>`_
* Wist *et al.* (2020) "Phenotypic and Genotypic Traits of Vancomycin-Resistant *Enterococci* from Healthy Food- Producing Animals"  *Microorganisms* `doi:10.3390/microorganisms8020261 <https://doi.org/10.3390/microorganisms8020261>`_
* Wu *et al.* (2020) "Toward a high-quality pan-genome landscape of *Bacillus subtilis* by removal of confounding strains" *Brief. Bioinf.* `doi:10.1093/bib/bbaa013 <https://doi.org/10.1093/bib/bbaa013>`_
* Yang *et al.* (2020) "Isolation and Characterization of the Novel Phages vB_VpS_BA3 and vB_VpS_CA8 for Lysing *Vibrio parahaemolyticus*" *Front. Microbiol.* `doi:10.3389/fmicb.2020.00259 <https://doi.org/10.3389/fmicb.2020.00259>`_
* Zayulina *et al.* "Complete Genome Sequence of a Hyperthermophilic Archaeon, *Thermosphaera* sp. Strain 3507, Isolated from a Chilean Hot Spring" *Micro. Res. Ann.* `doi:10.1128/MRA.01262-20 <https://doi.org/10.1128/MRA.01262-20>`_
* Zhang *et al.* (2020) "A novel bacterial thiosulfate oxidation pathway provides a new clue about the formation of zero-valent sulfur in deep sea." *ISME J.* `doi:10.1038/s41396-020-0684-5 <https://doi.org/10.1038/s41396-020-0684-5>`_
* Zhang *et al.* (2020) "*Deinococcus detaillensis* sp. nov., isolated from humus soil in Antarctica" *Arch. Microbiol.* `doi:10.1007/s00203-020-01920-0 <https://doi.org/10.1007/s00203-020-01920-0>`_
* Zhang *et al.* (2020) "Chloramphenicol biodegradation by enriched bacterial consortia and isolated strain *Sphingomonas* sp. CL5.1: The reconstruction of a novel biodegradation pathway" *Water Res.* `doi:10.1016/j.watres.2020.116397 <https://doi.org/10.1016/j.watres.2020.116397>`_
* Zheng *et al.* (2020) "Metagenomic Insight into Environmentally Challenged Methane-Fed Microbial Communities" *Microorganisms* `doi:10.3390/microorganisms8101614 <https://doi.org/10.3390/microorganisms8101614>`_
* Zhou *et al.* (2020) "Comparative analysis of *Lactobacillus gasseri* from Chinese subjects reveals a new species-level taxa" *BMC Genomics* `doi:10.1186/s12864-020-6527-y <https://doi.org/10.1186/s12864-020-6527-y>`_

^^^^
2019
^^^^

* Accetto & Avgustin (2019) "The diverse and extensive plant polysaccharide degradative apparatuses of the rumen and hindgut *Prevotella* species: A factor in their ubiquity?" *Syst. Appl. Microbiol.* `doi:j.syapm.2018.10.001 <https://doi.org/j.syapm.2018.10.001>`_
* Acevedo *et al.* (2019) "*Bacillus clarus* sp. nov. is a new *Bacillus cereus* group species isolated from soil" *BioRxiv* `doi:10.1101/508077 <https://doi.org/10.1101/508077>`_
* Alberoni *et al.* (2019) "*Bifidobacterium xylocopae* sp. nov. and *Bifidobacterium aemilianum *sp. nov., from the carpenter bee (*Xylocopa violacea*) digestive tract" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2018.11.005 <https://doi.org/10.1016/j.syapm.2018.11.005>`_
* Alex & Antunes (2019) "Whole-Genome Comparisons Among the Genus *Shewanella* Reveal the Enrichment of Genes Encoding Ankyrin-Repeats Containing Proteins in Sponge-Associated Bacteria" *Front. Microbiol.* `doi:10.3389/fmicb.2019.00005 <https://doi.org/10.3389/fmicb.2019.00005>`_
* Alex & Antunes (2019) "Comparative Genomics Reveals Metabolic Specificity of *Endozoicomonas* Isolated from a Marine Sponge and the Genomic Repertoire for Host-Bacteria Symbioses" *Microorganisms* `doi:10.3390/microorganisms7120635 <https://doi.org/10.3390/microorganisms7120635>`_
* Barnier *et al.* (2019) "Description of *Palleronia rufa* sp. nov., a biofilm-forming and AHL-producing *Rhodobacteraceae*, reclassification of *Hwanghaeicola aestuarii* as *Palleronia aestuarii* comb. nov., *Maribius pontilimi* as *Palleronia pontilimi* comb. nov., *Maribius salinus* as *Palleronia salina* comb. nov., *Maribius pelagius* as *Palleronia pelagia* comb. nov. and emended description of the genus *Palleronia*" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2019.126018 <https://doi.org/10.1016/j.syapm.2019.126018>`_
* Bayjanov *et al.* (2019) "Whole genome analysis of *Pandoraea* species strains from cystic fibrosis patients" *Future Microbiology* `doi:10.2217/fmb-2019-0038 <https://doi.org/10.2217/fmb-2019-0038>`_
* Botelho *et al.* (2019) "Combining sequencing approaches to fully resolve a carbapenemase-encoding megaplasmid in a Pseudomonas shirazica clinical strain" *Emerg. Microb. Inf.* `doi:10.1080/22221751.2019.1648182 <https://doi.org/10.1080/22221751.2019.1648182>`_
* Boukerb *et al.* (2019) "*Campylobacter armoricus* *sp. nov.*, a novel member of the *Campylobacter lari* group isolated from surface water and stools from humans with enteric infection" *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.003836 <https://doi.org/10.1099/ijsem.0.003836>`_
* Briand *et al.* (2019) "A rapid and simple method for assessing and representing genome sequence relatedness" *BioRxiv* `doi:10.1101/569640 <https://doi.org/10.1101/569640>`_
* Cho & Kwak (2019) "Evolution of Antibiotic Synthesis Gene Clusters in the *Streptomyces globisporus* TFH56, Isolated from Tomato Flower" *G3: Genes, Genomes, Genetics* `doi:10.1534/g3.119.400037  <https://dx.doi.org/10.1534/g3.119.400037>`_
* Ciok & Dziewit (2019) "Exploring the genome of Arctic *Psychrobacter* sp. DAB_AL32B and construction of novel *Psychrobacter*-specific cloning vectors of an increased carrying capacity" *Arch. Microbiol.* `doi:10.1007/s00203-018-1595-y <https://doi.org/10.1007/s00203-018-1595-y>`_
* D'Souza *et al.* (2019) "Spatiotemporal dynamics of multidrug resistant bacteria on intensive care unit surfaces" *Nat. Comm.* `doi:10.1038/s41467-019-12563-1 <https://doi.org/10.1038/s41467-019-12563-1>`_
* do Vale *et al.* (2019) "Draft Genome Sequences of Three Novel *Acinetobacter* Isolates from an Irish Commercial Pig Farm" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00919-19 <https://dx.doi.org/10.1128/MRA.00919-19>`_
* Doud *et al.* (2019) "Function-driven single-cell genomics uncovers cellulose-degrading bacteria from the rare biosphere" *ISME J.* `doi:10.1038/s41396-019-0557-y <https://doi.org/10.1038/s41396-019-0557-y>`_
* Du *et al.* (2019) "Characterization of a Linezolid- and Vancomycin-Resistant *Streptococcus suis* Isolate That Harbors optrA and vanG Operons" *Front. Microbiol.* `doi:10.3389/fmicb.2019.02026 <https://doi.org/10.3389/fmicb.2019.02026>`_
* Esposito *et al.* (2019) "Insights into the genome structure of four acetogenic bacteria with specific reference to the Wood–Ljungdahl pathway" *Microbiol. Open* `doi:10.1002/mbo3.938 <https://doi.org/10.1002/mbo3.938>`_
* Falagan *et al.* (2019) "Acidithiobacillus sulfuriphilus sp. nov.: an extremely acidophilic sulfur-oxidizing chemolithotroph isolated from a neutral pH environment" *Int. J. Syst. Evol. Micro.* `doi:0.1099/ijsem.0.003576 <https://doi.org/0.1099/ijsem.0.003576>`_
* Faoro *et al.* (2019) "Genome comparison between clinical and environmental strains of *Herbaspirillum seropedicae* reveals a potential new emerging bacterium adapted to human hosts" *BMC Genomics* `doi:10.1186/s12864-019-5982-9 <https://doi.org/10.1186/s12864-019-5982-9>`_
* Feng *et al.* (2019) "Complete genome sequence of *Hahella* sp. KA22, a prodigiosin-producing algicidal bacterium" *Marine Genomics* `doi:10.1016/j.margen.2019.04.003 <https://doi.org/10.1016/j.margen.2019.04.003>`_
* Gasparrini *et al.* (2019) "Metagenomic signatures of early life hospitalization and antibiotic treatment in the infant gut microbiota and resistome persist long after discharge" *Nature Microbiol.* `doi:10.1038/s41564-019-0550-2 <https://doi.org/10.1038/s41564-019-0550-2>`_
* Ghosh *et al.* (2019) "Reanalysis of *Lactobacillus paracasei* Lbs2 Strain and Large-Scale Comparative Genomics Places Many Strains into Their Correct Taxonomic Position" *Microorganisms* `doi:10.3390/microorganisms7110487 <https://doi.org/10.3390/microorganisms7110487>`_
* Hollensteiner *et al.* (2019) "Complete Genome Sequence of *Marinobacter* sp. Strain JH2, Isolated from Seawater of the Kiel Fjord" *Micro. Res. Ann.* `doi:10.1128/MRA.00596-19 <https://doi.org/10.1128/MRA.00596-19>`_
* Hornung *et al.* (2019) "An in silico survey of *Clostridioides difficile* extrachromosomal elements" *BioRxiv* `doi:10.1101/651539 <https://doi.org/10.1101/651539>`_
* Huang *et al.* (2019) "Genomic differences within the phylum Marinimicrobia: From waters to sediments in the Mariana Trench" *Marine Genomics* `doi:10.1016/j.margen.2019.100699 <https://doi.org/10.1016/j.margen.2019.100699>`_
* Ide *et al.* (2019) "Draft Genome Sequence of *Acidovorax* sp. Strain NB1, Isolated from a Nitrite-Oxidizing Enrichment Culture" *Micro. Res. Ann.* `doi:10.1128/MRA.00547-19 <https://doi.org/10.1128/MRA.00547-19>`_
* Jeong *et al.* (2019) "Chronicle of a Soil Bacterium: *Paenibacillus polymyxa* E681 as a Tiny Guardian of Plant and Human Health" *Front. Microbiol.* `doi:10.3389/fmicb.2019.00467 <https://doi.org/10.3389/fmicb.2019.00467>`_
* Kaminsky *et al.* (2019) "Genomic Analysis of γ-Hexachlorocyclohexane-Degrading *Sphingopyxis lindanitolerans* WS5A3p Strain in the Context of the Pangenome of *Sphingopyxis*" *Genes* `doi:0.3390/genes10090688 <https://doi.org/0.3390/genes10090688>`_
* Khan *et al.* (2019) "Genomic and physiological analyses reveal that extremely thermophilic *Caldicellulosiruptor changbaiensis* deploys unique cellulose attachment mechanisms" *BioRxiv* `doi:10.1101/622977 <https://doi.org/10.1101/622977>`_
* Kirmiz *et al.* (2019) "Comparative genomics guides elucidation of vitamin B12 biosynthesis in novel human associated *Akkermansia*" *BioRxiv* `doi:10.1101/587527 <https://doi.org/10.1101/587527>`_
* Kiu *et al.* (2019) "Genomic analysis on broiler-associated *Clostridium perfringens* strains and exploratory caecal microbiome investigation reveals key factors linked to poultry necrotic enteritis" *Animal Microbiome* `doi:10.1186/s42523-019-0015-1 <https://doi.org/10.1186/s42523-019-0015-1>`_
* Kiu *et al.* (2019) "Phylogenomic analysis of gastroenteritis-associated *Clostridium perfringens* in England and Wales over a 7-year period indicates distribution of clonal toxigenic strains in multiple outbreaks and extensive involvement of enterotoxin-encoding (CPE) plasmids" *Micro. Genom.* `doi:10.1099/mgen.0.000297 <https://doi.org/10.1099/mgen.0.000297>`_
* Lozada *et al.* (2019) "Phage vB_BmeM-Goe8 infecting *Bacillus megaterium* DSM319" *Arch. Virol.* `doi:10.1007/s00705-019-04513-5 <https://doi.org/10.1007/s00705-019-04513-5>`_
* Kochetkova *et al.* (2019) "*Tepidiforma bonchosmolovskayae* gen. nov., sp. nov., a moderately thermophilic *Chloroflexi* bacterium from a Chukotka hot spring (Arctic, Russia), representing a novel class, *Tepidiformia*, which includes the previously uncultivated lineage OLB14" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.003902 <https://doi.org/10.1099/ijsem.0.003902>`_
* Kovaleva *et al.* (2019) "*Tautonia sociabilis* gen. nov., sp. nov., a novel thermotolerant planctomycete, isolated from a 4000 m deep subterranean habitat" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.003467 <https://doi.org/10.1099/ijsem.0.003467>`_
* Labuda *et al.* (2019) "Bloodstream Infections With a Novel Nontuberculous Mycobacterium Involving 52 Outpatient Oncology Clinic Patients―Arkansas, 2018" *Clin. Inf. Dis.* `doi:10.1093/cid/ciz1120 <https://doi.org/10.1093/cid/ciz1120>`_
* Lan *et al.* (2019) "*Vogesella urethralis* *sp. nov.*, isolated from human urine, and emended descriptions of *Vogesella perlucida* and *Vogesella mureinivorans*" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.003802 <https://doi.org/10.1099/ijsem.0.003802>`_
* Lawson *et al.* (2019) "Breast milk-derived human milk oligosaccharides promote *Bifidobacterium* interactions within a single ecosystem" *ISME J.* `doi:0.1038/s41396-019-0553-2 <https://doi.org/0.1038/s41396-019-0553-2>`_
* Ma *et al.* (2019) "First report of *Dickeya fangzhongdai* causing soft rot of onion in New York State" *Plant Dis.* `doi:10.1094/PDIS-09-19-1940-PDN <https://doi.org/10.1094/PDIS-09-19-1940-PDN>`_
* Matteo-Estrada *et al.* (2019) "Phylogenomics Reveals Clear Cases of Misclassification and Genus-Wide Phylogenetic Markers for *Acinetobacter*" *Genome Biol. Evol.* `doi:10.1093/gbe/evz178 <https://doi.org/10.1093/gbe/evz178>`_
* McIntyre *et al.* (2019) "Single-molecule sequencing detection of N6-methyladenine in microbial reference materials" *Nat. Comm.* `doi:10.1038/s41467-019-08289-9 <https://doi.org/s41467-019-08289-9>`_
* Nordmann *et al.* (2019) "Complete genome sequence of the virus isolate vB_BthM-Goe5 infecting *Bacillus thuringiensis*" *Arch. Virol.* `doi:10.1007/s00705-019-04187-z <https://10.1007/s00705-019-04187-z>`_
* Paim *et al.* (2019) "Evaluation of niche adaptation features by genome data mining approach of *Escherichia coli* urinary and gastrointestinal strains" *PeerJ Preprints* `doi:10.7287/peerj.preprints.27720v1 <https://doi.org/10.7287/peerj.preprints.27720v1>`_
* Park *et al* (2019) "Complete genome sequence of acetate-producing *Klebsiella pneumoniae* L5-2 isolated from infant feces" *3Biotech* `doi:10.1007/s13205-019-1578-y <https://doi.org/10.1007/s13205-019-1578-y>`_
* Pedron & van Gijsegem (2019) "Diversity in the Bacterial Genus *Dickeya* Grouping Plant Pathogens and Waterways Isolates" *OBM Genetics* `doi:10.21926/obm.genet.1904098 <https://doi.org/10.21926/obm.genet.1904098>`_
* Portier *et al.* (2019) "Elevation of *Pectobacterium carotovorum* subsp. *odoriferum* to species level as *Pectobacterium odoriferum* sp. nov., proposal of *Pectobacterium brasiliense* sp. nov. and *Pectobacterium actinidiae* sp. nov., emended description of *Pectobacterium carotovorum* and description of *Pectobacterium versatile* sp. nov., isolated from streams and symptoms on diverse plants" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.003611 <https://doi.org/10.1099/ijsem.0.003611>`_
* Potter *et al.* (2019) "In Silico Analysis of *Gardnerella* Genomospecies Detected in the Setting of Bacterial Vaginosis" *Clin. Chem.* `doi:10.1373/clinchem.2019.305474 <https://doi.org/10.1373/clinchem.2019.305474>`_
* Reichler *et al.* (2019) "A century of gray: A genomic locus found in 2 distinct *Pseudomonas* spp. is associated with historical and contemporary color defects in dairy products worldwide" *J. Dairy Sci.* `doi:10.3168/jds.2018-16192 <https://doi.org/10.3168/jds.2018-16192>`_
* Royo-Llonch *et al.* "Ecological and functional capabilities of an uncultured *Kordia* sp" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2019.126045 <https://doi.org/10.1016/j.syapm.2019.126045>`_
* Ruiz *et al.* (2019) "Microbiota of human precolostrum and its potential role as a source of bacteria to the infant mouth" *Sci. Rep.* `doi:10.1038/s41598-019-42514-1 <https://doi.org/10.1038/s41598-019-42514-1>`_
* Sant'Anna *et al.* (2019) "Genomic metrics made easy: what to do and where to go in the new era of bacterial taxonomy" *Crit. Rev. Microbiol.* `doi:10.1080/1040841X.2019.1569587 <https://doi.org/10.1080/1040841X.2019.1569587>`_
* Schmuhl *et al.* (2019) "Comparative Transcriptomic Profiling of *Yersinia enterocolitica* O:3 and O:8 Reveals Major Expression Differences of Fitness- and Virulence-Relevant Genes Indicating Ecological Separation" *mSystems* `doi:10.1128/mSystems.00239-18 <https://doi.org/10.1128/mSystems.00239-18>`_
* Spirina *et al.* (2019) "Draft Genome Sequence of Microbacterium sp. Gd 4-13, Isolated from Gydanskiy Peninsula Permafrost Sediments of Marine Origin" *Microb. Res. Announce.* `doi:10.1128/MRA.00889-19 <https://doi.org/10.1128/MRA.00889-19>`_
* Stefanic *et al.* (2019) "Intra-species DNA exchange: *Bacillus subtilis* prefers sex with less related strains" *BioRxiv* `doi:10.1101/756569 <https://doi.org/10.1101/756569>`_
* Stevens *et al.* (2019) "Whole-genome-based phylogeny of *Bacillus cytotoxicus* reveals different clades within the species and provides clues on ecology and evolution" *Sci. Rep.* `doi:10.1038/s41598-018-36254-x <https://doi.org/10.1038/s41598-018-36254-x>`_
* Tanaka *et al.* (2019) "Draft Genome Sequences of *Enterococcus faecalis* Strains Isolated from Healthy Japanese Individuals" *Microb. Res. Announce.* `doi:10.1128/MRA.00832-19 <https://doi.org/10.1128/MRA.00832-19>`_
* Thorell *et al.* (2019) "Isolates from colonic spirochaetosis in humans show high genomic divergence and carry potential pathogenic features but are not detected by 16S amplicon sequencing using standard primers for the human microbiota" *BioRxiv* `doi:doi.org/10.1101/544502 <https://doi.org/doi.org/10.1101/544502>`_
* Tian *et al.* (2019) "LINbase: A Web service for genome-based identification of microbes as members of crowdsourced taxa" *BioRxiv* `doi:10.1101/752212 <https://doi.org/10.1101/752212>`_
* Tohno *et al.* (2019) "*Lactobacillus salitolerans* sp. nov., a novel lactic acid bacterium isolated from spent mushroom substrates" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.003224 <https://doi.org/10.1099/ijsem.0.003224>`_
* Vazquez-Campos *et al.* (2019) "Genomic insights into the Archaea inhabiting an Australian radioactive legacy site" *BioRxiv* `doi:10.1101/728089 <https://doi.org/10.1101/728089>`_
* Vincent *et al.* (2019) "A Mesophilic *Aeromona salmonicida* Strain Isolated from an Unsuspected Host, the Micratory Bird Pied Avocet" *Microorganisms* `doi:10.3390/microorganisms7120592 <https://doi.org/10.3390/microorganisms7120592>`_
* Vincent *et al.* (2019) "Investigation of the virulence and genomics of *Aeromonas salmonicida* strains isolated from human patients" *Inf. Genet. Evol.* `doi:10.1016/j.meegid.2018.11.019 <https://10.1016/j.meegid.2018.11.019>`_
* Vincent *et al.* (2019) "Revisiting the taxonomy and evolution of pathogenicity of the genus *Leptospira* through the prism of genomics" *PLoS Neg. Trop. Dis.* `doi:10.1371/journal.pntd.0007270 <https://doi.org/10.1371/journal.pntd.0007270>`_
* Wallner *et al.* (2019) "Genomic analyses of *Burkholderia cenocepacia* reveal multiple species with differential host-adaptation to plants and humans" *BMC Genomics* `doi:10.1186/s12864-019-6186-z <https://doi.org/10.1186/s12864-019-6186-z>`_
* Wang *et al.* (2019) "Occurrence of CTX-M-123-producing *Salmonella* Indiana in chicken carcasses: a new challenge for the poultry industry and food safety" *J. Antimicrob. Chemo.* `doi:10.1093/jac/dkz386 <https://doi.org/10.1093/jac/dkz386>`_
* Webster *et al.* (2019) "Genome Sequences of Two Choline-Utilizing Methanogenic Archaea, *Methanococcoides* spp., Isolated from Marine Sediments" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00342-19 <https://dx.doi.org/10.1128/MRA.00342-19>`_
* Webster *et al.* (2019) "The Genome Sequences of Three *Paraburkholderia* sp. Strains Isolated from Wood-Decay Fungi Reveal Them as Novel Species with Antimicrobial Biosynthetic Potential" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00778-19 <https://dx.doi.org/10.1128/MRA.00778-19>`_
* Wiegand *et al.* (2019) "Cultivation and functional characterization of 79 planctomycetes uncovers their unique biology" *Nat. Microbiol.* `doi:10.1038/s41564-019-0588-1 <https://doi.org/10.1038/s41564-019-0588-1>`_
* Wittouck *et al.* (2019) " A genome-based species taxonomy of the *Lactobacillus* genus complex" *mSystems* `doi:10.1128/mSystems.00264-19 <https://doi.org/10.1128/mSystems.00264-19>`_
* Yin *et al.* (2019) "A hybrid sub-lineage of *Listeria monocytogenes* comprising hypervirulent isolates" *Nat. Comm.* `doi:10.1038/s41467-019-12072-1 <https://doi.org/10.1038/s41467-019-12072-1>`_
* Yin *et al.* (2019) "Genetic Diversity of *Listeria monocytogenes* Isolates from Invasive Listeriosis in China" *Foodborne Path. Dis.* `doi:10.1089/fpd.2019.2693 <https://doi.org/10.1089/fpd.2019.2693>`_
* Zabel *et al.* (2019) "Novel Genes and Metabolite Trends in *Bifidobacterium longum* subsp. *infantis* Bi-26 Metabolism of Human Milk Oligosaccharide 2′-fucosyllactose" *Sci. Rep.* `doi:10.1038/s41598-019-43780-9 <https://doi.org/s41598-019-43780-9>`_
* Zakham *et al.* (2019) "Molecular diagnosis and enrichment culture identified a septic pseudoarthrosis due to an infection with *Erysipelatoclostridium ramosum*" *Int. J. Inf. Dis.* `doi:10.1016/j.ijid.2019.02.001 <https://doi.org/10.1016/j.ijid.2019.02.001>`_
* Zhu *et al.* (2019) "First Report of Integrative Conjugative Elements in *Riemerella anatipestifer* Isolates From Ducks in China" *Front. Vet. Sci.* `doi:10.3389/fvets.2019.00128 <https://doi.org/10.3389/fvets.2019.00128>`_
* Zhu *et al.* (2019) "Pan-genome analysis of *Riemerella anatipestifer* reveals its genomic diversity and acquired antibiotic resistance associated with genomic islands" *Func. Int. Genom* `doi:10.1007/s10142-019-00715-x <https://doi.org/10.1007/s10142-019-00715-x>`_

^^^^
2018
^^^^

* Alex & Antunes (2018) "Genus-wide comparison of *Pseudovibrio* bacterial genomes reveal diverse adaptations to different marine invertebrate hosts" *PLoS One* `doi:10.1371/journal.pone.0194368 <https://doi.org/10.1371/journal.pone.0194368>`_
* Beaton *et al.* (2018) "Community-led comparative genomic and phenotypic analysis of the aquaculture pathogen *Pseudomonas baetica* a390T sequenced by Ion semiconductor and Nanopore technologies" *FEMS Micro. Lett.* `doi:10.1093/femsle/fny069 <https://doi.org/10.1093/femsle/fny069>`_
* Bogema *et al.* (2018) "Analysis of *Theileria orientalis* draft genome sequences reveals potential species-level divergence of the Ikeda, Chitose and Buffeli genotypes" *BMC Genomics* `doi:10.1186/s12864-018-4701-2 <https://doi.org/10.1186/s12864-018-4701-2>`_
* Brand *et al.* (2018) "Niche Differentiation among Three Closely Related *Competibacteraceae* Clades at a Full-Scale Activated Sludge Wastewater Treatment Plant and Putative Linkages to Process Performance" *App. Env. Micro.* `doi:10.1128/AEM.02301-18 <https://doi.org/10.1128/AEM.02301-18>`_
* Bridel *et al.* (2018) "Comparative Genomics of *Tenacibaculum dicentrarchi* and “*Tenacibaculum finnmarkense*” Highlights Intricate Evolution of Fish-Pathogenic Species" *Genome Biol. Evol.* `doi:10.1093/gbe/evy020 <https://doi.org/10.1093/gbe/evy020>`_
* Carlos *et al.* (2018) "Substrate Shift Reveals Roles for Members of Bacterial Consortia in Degradation of Plant Cell Wall Polymers" *Front. Microbiol.* `doi:10.3389/fmicb.2018.00364 <https://doi.org/10.3389/fmicb.2018.00364>`_
* Covarrubias *et al.* (2018) "Occurrence, integrity and functionality of *Aca*ML1–like viruses infecting extreme acidophiles of the *Acidithiobacillus* species complex" *Res. Microbiol.* `doi:10.1016/j.resmic.2018.07.005 <http://doi.org/10.1016/j.resmic.2018.07.005>`_
* da Gama *et al.* (2018) "Taxonomic Repositioning of *Xanthomonas campestris* pv. *viticola* (Nayudu 1972) Dye 1978 as *Xanthomonas citri* pv. *viticola* (Nayudu 1972) Dye 1978 comb. nov. and Emendation of the Description of *Xanthomonas citri* pv. *anacardii* to Include Pigmented Isolates Pathogenic to Cashew Plant" *Phytopath.* `doi:10.1094/PHYTO-02-18-0037-R <https://doi.org/10.1094/PHYTO-02-18-0037-R>`_
* Ferretti *et al.* (2018) "Mother-to-Infant Microbial Transmission from Different Body Sites Shapes the Developing Infant Gut Microbiome" *Cell Host Microbe* `doi:10.1016/j.chom.2018.06.005 <https://doi.org/10.1016/j.chom.2018.06.005>`_
* Fontana *et al.* (2018) "Genetic Signatures of Dairy *Lactobacillus casei* Group" *Front. Microbiol.* `doi:10.3389/fmicb.2018.02611 <https://doi.org/10.3389/fmicb.2018.02611>`_
* Freschi *et al.* (2018) "The *Pseudomonas aeruginosa* Pan-Genome Provides New Insights on Its Population Structure, Horizontal Gene Transfer, and Pathogenicity" *Genome Biol. Evol.* `doi:10.1093/gbe/evy259 <https://doi.org/10.1093/gbe/evy259>`_
* Gillis *et al.* (2018) "Role of plasmid plasticity and mobile genetic elements in the entomopathogen *Bacillus thuringiensis* serovar *israelensis*" *FEMS Micro. Rev.* `doi:10.1093/femsre/fuy034 <https://doi.org/10.1093/femsre/fuy034>`_
* Gragna-Miraglia *et al.* (2018) "Phylogenomics picks out the par excellence markers for species phylogeny in the genus *Staphylococcus*" *PeerJ* `doi:10.7717/peerj.5839 <https://doi.org/10.7717/peerj.5839>`_
* Hubbard *et al.* (2018) "Comparison of the first whole genome sequence of ‘*Haemophilus quentini*’ with two new strains of ‘*Haemophilus quentini*’ and other species of *Haemophilus*" *Genome* `doi:10.1139/gen-2017-0195 <https://doi.org/10.1139/gen-2017-0195>`_
* Issotta *et al.* (2018) "Insights into the biology of acidophilic members of the *Acidiferrobacteraceae* family derived from comparative genomic analyses" *Res. Microbiol.* `doi:10.1016/j.resmic.2018.08.001 <https://doi.org/10.1016/j.resmic.2018.08.001>`_
* Jangam *et al.* (2018) "Draft Genome Sequence of *Vibrio parahaemolyticus* Strain VP14, Isolated from a *Penaeus vannamei* Culture Farm" *Micro. Res. Ann.* `doi:10.1128/genomeA.00149-18 <https://10.1128/genomeA.00149-18>`_
* Jarett *et al.* (2018) "Single-cell genomics of co-sorted *Nanoarchaeota* suggests novel putative host associations and diversification of proteins involved in symbiosis" *Microbiome* `doi:10.1186/s40168-018-0539-8 <https://doi.org/10.1186/s40168-018-0539-8>`_
* Jung *et al.* (2018) "Complete genome sequence of *Bifidobacterium choerinum* FMB-1, a resistant starch-degrading bacterium" *J. Biotech.* `doi:10.1016/j.jbiotec.2018.03.009 <https://doi.org/10.1016/j.jbiotec.2018.03.009>`_
* Lazarte *et al.* (2018) "*Bacillus wiedmannii* biovar *thuringiensis*: A Specialized Mosquitocidal Pathogen with Plasmids from Diverse Origins" *Genome Biol. Evol.* `doi:10.1093/gbe/evy211 <https://doi.org/10.1093/gbe/evy211>`_
* Li *et al.* (2018) "A Novel Strategy for Detecting Recent Horizontal Gene Transfer and Its Application to *Rhizobium* Strains" *Front. Microbiol.* `doi:10.3389/fmicb.2018.00973 <https://dx.doi.org/10.3389/fmicb.2018.00973>`_
* Lima *et al.* "Genome sequencing and functional characterization of the non-pathogenic *Klebsiella pneumoniae* KpGe bacteria* *Microbes Inf.* `doi:10.1016/j.micinf.2018.04.001 <https://doi.org/10.1016/j.micinf.2018.04.001>`_
* McCann *et al.* (2018) "Viromes of one year old infants reveal the impact of birth mode on microbiome diversity" *PeerJ* `doi:10.7717/peerj.4694 <https://doi.org/10.7717/peerj.4694>`_
* Morales-Covarrubias (2018) "*Streptococcus penaeicida* sp. nov., isolated from a diseased farmed Pacific white shrimp (*Penaeus vannamei*)" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.002693 <https://doi.org/10.1099/ijsem.0.002693>`_
* Munoz-Villagran *et al.* (2018) "Comparative genomic analysis of a new tellurite-resistant *Psychrobacter* strain isolated from the Antarctic Peninsula" *PeerJ* `doi:10.7717/peerj.4402 <https://doi.org/10.7717/peerj.4402>`_
* Nascimento *et al.* (2018) "From plants to nematodes: *Serratia grimesii* BXF1 genome reveals an adaptation to the modulation of multi-species interactions" *Microb. Genom.* `doi:10.1099/mgen.0.000178 <https://doi.org/10.1099/mgen.0.000178>`_
* Orr *et al.* (2018) "De novo assembly of the *Pasteuria penetrans* genome reveals high plasticity, host dependency, and BclA-like collagens" *BioRxiv* `doi:10.1101/485748 <https://doi.org/10.1101/485748>`_
* Pinto *et al.* (2018) "Draft Genome Sequences of Novel *Pseudomonas*, *Flavobacterium*, and *Sediminibacterium* Strains from a Freshwater Ecosystem" *Micro. Res. Ann.* `doi:10.1128/genomeA.00009-18 <https://doi.org/10.1128/genomeA.00009-18>`_
* Potter *et al.* (2018) "Population Structure, Antibiotic Resistance, and Uropathogenicity of *Klebsiella variicola*" *mBio* `doi:10.1128/mBio.02481-18 <https://doi.org/10.1128/mBio.02481-18>`_
* Potter *et al.* (2018) "*Superficieibacter electus* gen. nov., sp. nov., an Extended-Spectrum β-Lactamase Possessing Member of the Enterobacteriaceae Family, Isolated From Intensive Care Unit Surfaces" *Front. Microbiol.* `doi:10.3389/fmicb.2018.01629 <https://doi.org/10.3389/fmicb.2018.01629>`_
* Samad *et al.* (2017) "Comparative genome analysis of the vineyard weed endophyte *Pseudomonas viridiflava* CDRTc14 showing selective herbicidal activity" *Sci. Rep.* `doi:10.1038/s41598-017-16495-y <https://doi.org/10.1038/s41598-017-16495-y>`_
* Sant'Anna *et al.* (2018) "Genome-based reclassification of *Paenibacillus dauci* as a later heterotypic synonym of *Paenibacillus shenyangensis*" *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.003127 <https://10.1099/ijsem.0.003127>`_
* Schilling *et al.* (2018) "Genomic Analysis of the Recent Viral Isolate vB_BthP-Goe4 Reveals Increased Diversity of φ29-Like Phages" *Viruses* `doi:10.3390/v10110624 <https://doi.org/10.3390/v10110624>`_
* Stevens *et al.* (2018) "Massive Diversity in Whole-Genome Sequences of *Streptococcus suis* Strains from Infected Pigs in Switzerland" *Microbiol. Res. Ann.* `doi:10.1128/MRA.01656-18 <https://dx.doi.org/10.1128/MRA.01656-18>`_
* Tanizawa *et al.* (2018) "Lactobacillus paragasseri sp. nov., a sister taxon of Lactobacillus gasseri, based on whole-genome sequence analyses" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.003020 <https://doi.org/10.1099/ijsem.0.003020>`_
* Vincent & Charette (2018) "Completion of genome of *Aeromonas salmonicida* subsp. *salmonicida* 01-B526 reveals how sequencing technologies can influence sequence quality and result interpretations" *New Microb. New Inf.* `doi:10.1016/j.nmni.2018.05.007 <https://doi.org/10.1016/j.nmni.2018.05.007>`_
* Wilhelm (2018) "Following the terrestrial tracks of *Caulobacter* - redefining the ecology of a reputed aquatic oligotroph" *ISME J* `doi:10.1038/s41396-018-0257-z <https://doi.org/10.1038/s41396-018-0257-z>`_
* Wittwer *et al.* (2018) "Population Genomics of *Francisella tularensis* subsp. *holarctica* and its Implication on the Eco-Epidemiology of Tularemia in Switzerland" *Front. Cell. Inf. Microbiol.* `doi:10.3389/fcimb.2018.00089 <https://doi.org/10.3389/fcimb.2018.00089>`_
* Zhang *et al.* (2018) "Draft Genome Sequence of *Komagataeibacter maltaceti* LMG 1529T, a Vinegar-Producing Acetic Acid Bacterium Isolated from Malt Vinegar Brewery Acetifiers" *Micro. Res. Ann.* `doi:10.1128/genomeA.00330-18 <https://doi.org/10.1128/genomeA.00330-18>`_

^^^^
2017
^^^^

* Anderson *et al.* (2017) "Genomic variation in microbial populations inhabiting the marine subseafloor at deep-sea hydrothermal vents" *Nat. Comm.* `doi:10.1038/s41467-017-01228-6 <https://doi.org/10.1038/s41467-017-01228-6>`_
* Ding *et al.* (2017) "Loss of the ssrA genome island led to partial debromination in the PBDE respiring *Dehalococcoides mccartyi* strain GY50" *Env. Micro.* `doi:10.1111/1462-2920.13817 <https://doi.org/10.1111/1462-2920.13817>`_
* Edgington *et al.* (2017) "Genome Sequences of Chancellor, Mitti, and Wintermute, Three Subcluster K4 Phages Isolated Using *Mycobacterium smegmatis* mc^{2}155" *Microbiol. Res. Ann.* `doi:10.1128/genomeA.01070-17 <https://doi.org/10.1128/genomeA.01070-17>`_
* Esposito *et al.* (2017) "Evolution of *Stenotrophomonas maltophilia* in Cystic Fibrosis Lung over Chronic Infection: A Genomic and Phenotypic Population Study" *Front. Microbiol.* `doi:10.3389/fmicb.2017.01590 <https://10.3389/fmicb.2017.01590>`_
* Jeukens *et al.* (2017) "A Pan-Genomic Approach to Understand the Basis of Host Adaptation in *Achromobacter*" *Genome Biol. Evol.* `doi:10.1093/gbe/evx061 <https://doi.org/10.1093/gbe/evx061>`_
* Ke *et al.* (2017) "Comparative genomics of *Vibrio campbellii* strains and core species of the *Vibrio Harveyi* clade" *Sci. Rep.* `doi:10.1038/srep41394 <https://doi.org/10.1038/srep41394>`_
* Kumar *et al.* (2017) "Draft Genome Sequence of the Luminescent Strain *Vibrio campbellii* LB102, Isolated from a Black Tiger Shrimp (*Penaeus monodon*) Broodstock Rearing System" *Micro. Res. Ann.* `doi:10.1128/genomeA.00342-17 <https://doi.org/10.1128/genomeA.00342-17>`_
* Pelve *et al.* (2017) "Bacterial Succession on Sinking Particles in the Ocean's Interior" *Front. Microbiol.* `doi:10.3389/fmicb.2017.02269 <https://doi.org/10.3389/fmicb.2017.02269>`_
* Poehlein *et al.* (2017) "Microbial solvent formation revisited by comparative genome analysis" *Biotech. Biofuels* `doi:10.1186/s13068-017-0742-z <https://doi.org/10.1186/s13068-017-0742-z>`_
* Ruiz-Valdeviezo *et al.* (2017) "Complete Genome Sequence of a Novel Nonnodulating *Rhizobium* Species Isolated from *Agave americana* L. Rhizosphere" *Micro. Res. Ann.* `doi:10.1128/genomeA.01280-17 <https://doi.org/10.1128/genomeA.01280-17>`_
* Tada *et al.* (2017) "Revealing the genomic differences between two subgroups in *Lactobacillus gasseri*" *Biosci. Microb. Food Health* `doi:10.12938/bmfh.17-006 <https://doi.org/10.12938/bmfh.17-006>`_
* Tanizawa *et al.* (2017) "Genomic characterization reconfirms the taxonomic status of *Lactobacillus parakefiri*" *Biosci. Microb. Food Health* `doi:10.12938/bmfh.16-026 <https://doi.org/10.12938/bmfh.16-026>`_
* Tohno *et al.* (2017) "*Lactobacillus silagincola* sp. nov. and *Lactobacillus pentosiphilus* sp. nov., isolated from silage" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.002196 <https://doi.org/10.1099/ijsem.0.002196>`_
* Vincent *et al.* (2017) "Study of mesophilic *Aeromonas salmonicida* A527 strain sheds light on the species’ lifestyles and taxonomic dilemma" *FEMS Micro. Lett.* `doi:10.1093/femsle/fnx239 <https://doi.org/10.1093/femsle/fnx239>`_
* Vollmers *et al.* (2017) "Untangling Genomes of Novel *Planctomycetal* and *Verrucomicrobial* Species from Monterey Bay Kelp Forest Metagenomes by Refined Binning" *Front. Microbiol.* `doi:10.3389/fmicb.2017.00472 <https://doi.org/10.3389/fmicb.2017.00472>`_
* Wang *et al.* (2017) "Genomic sequence of 'Candidatus *Liberibacter solanacearum*' haplotype C and its comparison with haplotype A and B genomes" *PLoS One* `doi:10.1371/journal.pone.0171531 <https://doi.org/10.1371/journal.pone.0171531>`_

^^^^
2016
^^^^

* Burstein *et al.* (2016) "New CRISPR–Cas systems from uncultivated microbes" *Nature* `doi:10.1038/nature21059 <https://doi.org/10.1038/nature21059>`_
* Gupta *et al.* (2016) "Comparative genomic analysis of novel *Acinetobacter* symbionts: A combined systems biology and genomics approach" *Sci. Rep.* `doi:10.1038/srep29043 <https://doi.org/srep29043>`_
* Haack *et al.* (2016) "Molecular Keys to the *Janthinobacterium* and *Duganella* spp. Interaction with the Plant Pathogen *Fusarium graminearum*" *Front. Microbiol.* `doi:10.3389/fmicb.2016.01668 <https://dx.doi.org/10.3389/fmicb.2016.01668>`_
* Maeno *et al.* (2016) "Genomic characterization of a fructophilic bee symbiont *Lactobacillus kunkeei* reveals its niche-specific adaptation" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2016.09.006 <https://doi.org/10.1016/j.syapm.2016.09.006>`_
* Pritchard *et al.* (2016) "Genomics and taxonomy in diagnostics for food security: soft-rotting enterobacterial plant pathogens" *Anal. Methods* `doi:10.1039/C5AY02550H <https://doi.org/10.1039/C5AY02550H>`_
* Rodriguez-Rojas *et al.* (2016) "Draft Genome Sequence of a Multi-Metal Resistant Bacterium *Pseudomonas putida* ATH-43 Isolated from Greenwich Island, Antarctica" *Front. Microbiol.* `doi:10.3389/fmicb.2016.01777 <https://doi.org/10.3389/fmicb.2016.01777>`_
* Tanizawa *et al.* (2016) "DFAST and DAGA: web-based integrated genome annotation tools and resources" *Biosci. Microb. Food Health* `doi:10.12938/bmfh.16-003 <https://doi.org/10.12938/bmfh.16-003>`_
* Zheng *et al.* (2016) "Metabolism of Toxic Sugars by Strains of the Bee Gut Symbiont *Gilliamella apicola*" *mBio* `doi:10.1128/mBio.01326-16 <https://doi.org/10.1128/mBio.01326-16>`_


.. _10.1039/C5AY02550H: https://dx.doi.org/10.1039/C5AY02550H
